TeachingRecordSystem/src/TeachingRecordSystem.Core/TrnRequestHelper.cs

```diff
@@ -34,52 +34,70 @@ public string GetAccessYourTeachingQualificationsLink(string trnToken) =>
         var dbTrnRequest = await dbContext.TrnRequests.SingleOrDefaultAsync(r => r.ClientId == applicationUserId.ToString() && r.RequestId == requestId);
         var metadata = await dbContext.TrnRequestMetadata.SingleOrDefaultAsync(r => r.ApplicationUserId == applicationUserId && r.RequestId == requestId);
 
+        Guid contactId;
+
         if (metadata is null)
         {
             return null;
         }
 
-        if (dbTrnRequest is not null)
+        if (metadata.ResolvedPersonId is Guid personId)
+        {
+            contactId = personId;
+        }
+        else if (dbTrnRequest is not null)
         {
-            var contact = await GetContactAsync(dbTrnRequest.TeacherId);
-            return new(applicationUserId, contact, metadata);
+            contactId = dbTrnRequest.TeacherId;
         }
+        else if (await getContactByTrnRequestIdTask is Contact trnRequestContact)
+        {
+            contactId = trnRequestContact.Id;
+        }
+        else
+        {
+            return null;
+        }
+
+        var contact = await crmQueryDispatcher.ExecuteQueryAsync(
+            new GetContactWithMergeResolutionQuery(
+                contactId,
+                new ColumnSet(
+                    Contact.Fields.dfeta_TRN,
+                    Contact.Fields.FirstName,
+                    Contact.Fields.MiddleName,
+                    Contact.Fields.LastName,
+                    Contact.Fields.dfeta_StatedFirstName,
+                    Contact.Fields.dfeta_StatedMiddleName,
+                    Contact.Fields.dfeta_StatedLastName,
+                    Contact.Fields.EMailAddress1,
+                    Contact.Fields.dfeta_NINumber,
+                    Contact.Fields.BirthDate,
+                    Contact.Fields.Merged,
+                    Contact.Fields.MasterId,
+                    Contact.Fields.dfeta_SlugId,
+                    Contact.Fields.dfeta_QTSDate,
+                    Contact.Fields.dfeta_TrnToken)));
+
+        var result = new GetTrnRequestResult(applicationUserId, contact, metadata);
 
-        if (await getContactByTrnRequestIdTask is Contact trnRequestContact)
+        // If the request is Completed, ensure we've got a TRN and TRN token assigned to metadata
+        if (result.IsCompleted)
         {
-            var contact = await GetContactAsync(trnRequestContact.Id);
-            return new(applicationUserId, contact, metadata);
+            if (metadata.TrnToken is null && metadata.EmailAddress is not null)
+            {
+                metadata.TrnToken = await CreateTrnTokenAsync(result.Trn, metadata.EmailAddress);
+            }
+
+            metadata.ResolvedPersonId ??= result.Contact.Id;
+
+            await dbContext.SaveChangesAsync();
```

Comment:
nitpicking I guess but this method is a "Get" and it seems to be updating the DB

Reply:
It's true, but the primary function is still retrieving a `TrnRequestResult`